brushfire-core/src/main/scala/com/stripe/brushfire/TDigest.scala

```diff
@@ -0,0 +1,78 @@
+package com.stripe.brushfire
+
+import com.tdunning.math.stats.TDigest
+import com.twitter.algebird.{Monoid, Semigroup}
+
+private [this] case object TDigestSemigroup extends Semigroup[TDigest] {
+  override def plus(l: TDigest, r: TDigest): TDigest = {
+    val td = TDigest.createDigest(math.max(l.compression(), r.compression()))
+    td.add(l)
+    td.add(r)
+    td
+  }
+
+  override def sumOption(iter: TraversableOnce[TDigest]): Option[TDigest] = {
+    iter.foldLeft(None: Option[TDigest]) {
+      case (None, el) =>
+        val td = TDigest.createDigest(el.compression())
+        td.add(el)
+        Some(td)
+
+      case (f@Some(acc), el) =>
+        acc.add(el)
+        f
+    }
+  }
+}
+
+object TDigestSplitter {
+  private def splitDists[L](q: Double)(target: L, digest: TDigest): (Map[L, Long], Map[L, Long]) = {
+    // the cumulative density reported is sometimes outside of [0,1] so we need to bound it. otherwise
+    // the estimated target distribution will be far from realistic... leading the Evaluator to
+    // frequently pick a suboptimal split point
+    val leftDist = (digest.cdf(q).max(0.0).min(1.0) * digest.size().toDouble).toLong
+    val rightDist = (digest.size - leftDist).max(0L)
+    val left = if (leftDist > 0L) Map(target -> leftDist) else Map.empty[L, Long]
+    val right = if (rightDist > 0L) Map(target -> rightDist) else Map.empty[L, Long]
+    (left, right)
+  }
+}
+
+case class TDigestSplitter[L](k: Int = 25, compression: Double = 100.0) extends Splitter[Double, Map[L, Long]] {
+  override type S = Map[L, TDigest]
+
+  override def split(parent: Map[L, Long], stats: S): Iterable[Split[Double, Map[L, Long]]] = {
+    implicit val tds = TDigestSemigroup
+    import TDigestSplitter.splitDists
+
+    val splits = for {
+      // merge the statistics from all targets
+      merged <- Semigroup.sumOption(stats.valuesIterator).toSeq
+      i <- 1 to k
+      q = merged.quantile(i.toDouble / k.toDouble).max(0.0).min(merged.size().toDouble)
+      (leftDist, rightDist) = Monoid.sum(stats.map(Function.tupled(splitDists(q))))
+      if leftDist.nonEmpty || rightDist.nonEmpty
+    } yield {
+      BinarySplit(LessThan(q), leftDist, rightDist)
```

Comment:
You had asked earlier about the `Group.minus` thing. I think it's fine to ignore for now but let me explain: the joint distribution, `S`, is only going to include instances who have a value for the feature we're splitting on. But there may be some instances that would have made it to the node we're splitting, but *don't* have that feature. We have the distribution for the parent node, so if we want, and if our target is a `Group`, we can figure out how many instances must be missing this feature by doing the subtraction (and choose to include them on one or the other sides of the split).

However, that's effectively hard-coding the missing-feature traversal strategy during training (something we *don't* do anymore in evaluation). So I'm not even sure if it's the right thing, or if we should default to just dropping them.

The one place we *need* it is in the sparse boolean splitter, where (for efficiency) we expect a one-hot encoding where ie the feature only appears if at all if it's true and the falses are all missing.